app/src/main/report/components/Sidebar/PublishSwitch.tsx

```diff
@@ -33,30 +58,57 @@ export class PublishSwitchComponent extends React.Component<Props, undefined> {
         }
     }
 
+    getPublishVerb(){
+        return this.props.published ? 'unpublish' : 'publish';
+    }
+
     render() {
         return <div className="pt-3" id={"publish"}>
+                <PublishSwitchModal show={this.state.showModal} publishVerb={this.getPublishVerb()} onClose={this.hideModal}
+                    onConfirm={this.onConfirm}/>
                 <Toggle
                     onClick={this.onToggle}
                     on={<span>Published</span>}
                     off={<span>Internal</span>}
                     offstyle="internal"
                     onstyle={"published"}
                     active={this.props.published}/>
-                <UncontrolledTooltip target={"publish"}>
-                    Publish functionality coming soon</UncontrolledTooltip>
             </div>
     }
 }
 
-const mapStateToProps = (state: ReportAppState, props: PublicProps): Partial<Props> => {
+export class PublishSwitchModal extends React.Component<PublishSwitchModalProps, undefined> {
```

Comment:
I initially tried to make the modal using the bootstrap modal support via reactstrap, but found that Modal required react v16. Attempting to upgrade using npm seemed to break all sorts of things so I've gone with simple handrolled solution for now.